### ISSUE TITLE
Add seeking tests

### ISSUE THREAD

**Issue by thiblahute** [pitivi/gst-editing-services#30]

**Comment by vliaskov** [comment]
I have an early test patch here:
https://github.com/vliaskov/PitiviGes/commit/099ecf013aebaa6d9ba600ab5f56f3d9d18a0740

the patch adds a seek argument to test_timeline_with_profile. The seek value is used in gst_element_seek
So far seeking_playback only seeks in the pre-existing timeline already defined there (two 1 second clips appended to each other, the first with an effect)

if i do multiple seeks i.e. calling seeking_playback with different positions) i get some timeout expirations. Is this timeout a check-framework variable? (i 've seen timeout messages for other non-seek intgeration tests as well).
ges/integration.c:289:E:render:test_seeking_playback_webm:0: (after this point) Test timeout expired

Seeking in a sub-seconds granularity within this 2-second timeline seems to work with .webm format. but .ogg doesn't seem accurate. (I am also using GST_SEEK_FLAG_ACCURATE but didn't see a difference with or without). Should we also generate longer timelines and/or testfiles for seeking tests?

**Comment by MathieuDuponchelle** [comment]
The timeout is indeed a variable of the check framework, you can disable it but a priori it means something has failed, as our tests don't exceed the timeout value. If one test legitimately takes more time we can change that timeout as well.

I have patches for oggdemux waiting to be merged on gst-plugins-base, that make seeking accurate back again:
https://bugzilla.gnome.org/show_bug.cgi?id=700537

As for your patch, cool, it's great you took so little time to have something functionnal !

I imagine the next step would be to have something like gnonlin : 
https://github.com/pitivi/gnonlin/blob/master/tests/check/gnl/seek.c

Basically there is a list of seeks, the difference for us is that gnonlin doesn't actually display the result, and perform these seeks sequentially without interval, relying on a pad probe to make sure seeks succeeded.

I think we want to think of a way to perform multiple seeks one after another with an interval, I'd like to know what @thiblahute  thinks about that.

Having longer timelines could also be a good idea, I'm not sure yet.

Thanks a lot !

**Comment by vliaskov** [comment]
thanks for pointing out the gnonlin tests link and for the oggdemux buglink.

Can you explain the "with / without interval" concept? i.e. the difference between the gnonlin tests and what we want to have? By "interval" do you mean letting the pipeline resume between seek events, as opposed to batching the seeks the way gnonling seek test does?

**Comment by thiblahute** [comment]
So the idea would be to do something a bit smarter for the GES tests which is, adding a seeking_position or somthing like that to the SeekInfo structure and execut the seek at tha position.

In the end we will need to get the position of the pipeline in frequent interval, the same way we are currently doing it in ges-launch: https://github.com/pitivi/gst-editing-services/blob/master/tools/ges-launch.c#L387 and execute the seek at the proper time (you will need some tolerance interval).

Then using that same get_position callback, you should be able to check that after the seek the position is the one you just seeked to keeping any reference to the seeked value, but as the seeks are async, you will need to wai for ASYNC_DONE on the bus, and actually check the new position after that, so it will basically look like:

``` c

static GstClockTime seeked_position = GST_CLOCK_TIME_NONE;
static gboolean got_async_done = FALSE;

   /* On the bus callback */
   case ASYNC_DONE:
      if (GST_CLOCK_TIME_IS_VALID (seeked_position)) {
          got_async_done = TRUE;
      }
...

  /* position callback */
  position = gst_element_get_position(pipeline);
  for (tmp = wanted_seeks; tmp; tmp = tmp->next) {
    SeekInfo *sinfo = (SeekInfo) tmp->data;

    if (sinfo->seeking_position +/-== position) {
      fail_if (GST_CLOCK_TIME_IS_VALID (seeked_position)); /* it ould mean the previous seek was not properly executed */
      seeked_position = sinfo->position;
      fail_unless (gst_element_seek (pipeline, sinfo->position));
      got_async_done = FALSE
      break;
     }
  }
  if  (got_async_done) {
     fail_unless (position +/-== seeked_position);
     seeked_position = GST_CLOCK_TIME_NONE;
      got_async_done = FALSE
  }
```

Come and ask on IRC if you need more info :)

**Comment by vliaskov** [comment]
thanks for the details. I have another version that implements the get_position and a global list of seekinfo structs. The current _seeking_playback example now performs 3 seeks using this infrastructure.

https://github.com/vliaskov/PitiviGes/commit/7383969a8d94d3c3e817a7fd074d60e8089e50c0

get_position is called every 50msec, and tolerance interval is also 50 msec, but these are just example values.

I only tried with webm format so far. Multiple forward seeks work, but backward seeks don't work. Not sure if it's the format, an issue with gst_element_seek_simple vs gst_element_seek (although both don't work i believe) or some other bug.

**Comment by MathieuDuponchelle** [comment]
That's just great now !

seek vs. seek_simple should be equivalent.

I merged your patch on my integration branch, and pushed it on 
https://github.com/pitivi/gst-editing-services/commits/integration .

previous integration commits have been pushed on master, no doubt this commit will follow shortly :)

Thanks a lot, this was an essential thing to do.

**Comment by MathieuDuponchelle** [comment]
Also, I tried with MOV files and the test timed out, which is interesting, and exactly the reason why we needed these tests.

**Comment by thiblahute** [comment]
I made a few comments on https://github.com/vliaskov/PitiviGes/commit/7383969a8d94d3c3e817a7fd074d60e8089e50c0 please fix them and then I should be able to merge upstream :)

**Comment by thiblahute** [comment]
BTW, please base you fixes on https://github.com/pitivi/gst-editing-services/commits/integration

And thanks for that :)

**Comment by vliaskov** [comment]
a patch with some fixes is here:
https://github.com/vliaskov/PitiviGes/commit/9dcf26cff0b8ca925c161f3427929f2a520d2dd9

**Comment by thiblahute** [comment]
Ok, this is now merged, thanks a lot vliaskov, and we are looking forward working more with you :)

**Comment by MathieuDuponchelle** [comment]
Reopening because we also want seek tests while PAUSED, which is something we often do in pitivi.

**Comment by vliaskov** [comment]
is something as simple as this adequate?
https://github.com/vliaskov/PitiviGes/commit/96811cbc820117a119c1cd909bc859a9195f9d87

**Comment by thiblahute** [comment]
It is a good test yes.

I would also like to have tests where we do not go to playing, but just
seek while PAUSED and wait for seek_done in between.

**Comment by vliaskov** [comment]
2nd paused-seek test (on top of first) performs seeks on a PAUSED pipeline:
https://github.com/vliaskov/PitiviGes/commit/1ba35adfe48f51a3dbcc5408c8bf1113bb12188f

**Comment by thiblahute** [comment]
Great, this is now all done :)
